_posts/2020-10-24-text-layout.md

```diff
@@ -0,0 +1,217 @@
+
+## Implementations to study
+
+While I still feel a need for a solid, high-level, cross-platform text layout engine, there are good implementations to study. In open source, on of my favorites (though I am biased), is the Android text stack, based on [Minikin] for its lower levels. It is fairly capable and efficient, and also makes a concerted effort to get "all of Unicode" right, including emoji. It is also reasonably simple and the code is accessible.
+
+While not open source, [DirectWrite] is also well worth study, as it is without question one of the most capable engines, supporting Word and the previous iteration of Edge before it was abandonded in favor of Chromium. Note that there is a [proposal][DWriteCore proposal] for a cross-platform implementation and also potentially to take it open-source. If that were to happen, it would be something of a game changer.
+
+### Android
+
+Paragraph and style segmentation (with BiDi) is done at higher levels, in [Layout.java] and [StaticLayout.java]. At that point, runs are handed to [Minikin] for lower-level processing. Most of the rest of the hierarchy is in Layout.cpp, and ultimately shaping is done by HarfBuzz.
+
+Minikin also contains a sophisticated line breaking implementation, including Knuth-Plass style optimized breaking.
+
+Android deals with shaping boundaries by using heuristics to further segment the text to implied word boundaries (which are also used as the grain for layout cache). If a font does shaping across these boundaries, the shaping context is simply lost. This is a reasonable compromise, especially in mobile, as results are always consistent, ie the width for measurement never mismatches the width for layout. And none of the fonts in the system stack have exotic behavior such as shaping across spaces.
+
+Android does base its itemization on cmap coverage, and builds sophisticated bitmap structures for fast queries. As such, it can get normalization issues wrong, but overall this seems like a reasonable compromise. In particular, most of the time you'll run into normalization issues is with Latin and the combining diacritical marks, both of which are supplied by Roboto, which in turn has massive Unicode coverage (and thus less need to rely on normalization logic). But with custom fonts, handling may be less than ideal, resulting in more fallback to Roboto than might actually be needed.
+
+Note that Minikin was also the starting point for [libTxt](https://github.com/flutter/flutter/issues/11092), the text layout library used in Flutter.
+
+## DirectWrite
```

Comment:
I *think* you intended this to be ### rather than ## - the other implementation headers are all ###